### PR TITLE
[NCCL][CUDA][CUDA Graphs] Flush enqueued work before starting a graph capture

### PR DESCRIPTION
**Pull request by eqy**
#103503 addresses the situation where additional work is enqueued for the NCCL watchdog to poll during a graph capture---something we want to avoid as the subsequent polling will query an event and crash the capture.

However, there is currently no check that there was not work _already_ enqueued for the watchdog to poll. If there was already work that was enqueued and not cleaned up before the start of a graph capture, then we run into a similar problem where the event query will crash the capture. We've observed this causing crashes on several workloads, although it is somewhat flaky (if the watchdog happens to poll just before the graph capture and cleanup, then we dodge the crash).

This is a bit of a tricky issue as it involves making sure that no process group has enqueued work at the start of a capture, and as such the simplest solution is to add a bit of global state to track all process groups. This PR forces the start of the graph capture to wait until all enqueued work is completed and cleaned up or times out.

I did consider the alternative of simply having the watchdog skip cleanup if we detect that we are in the middle of a graph capture, but I think deferring the cleanup until later could result in false positive timeouts (e.g., if we defer cleanup on work that has completed long ago, checking timers after the graph capture could yield a "timeout").

CC @Aidyn-A 
@bottler @kwen2501 @ptrblck 

cc @ptrblck @mcarilli @ezyang

### CODE REVIEW

**Comment by kwen2501** [comment]
Thanks @eqy.
I wonder if CUDA Graph has a mode that would allow a side thread to concurrently query an event that's created **before** the capture sequence? https://github.com/pytorch/pytorch/pull/103503 fails because the event is created inside the capture sequence so I agree that the query is undefined. Would `cudaStreamCaptureModeThreadLocal` or `cudaStreamCaptureModeRelaxed` help here?

**Comment by eqy** [comment]
I have a fairly reliable repro for this bug so I'll check if `cudaStreamCaptureModeRelaxed` also addresses the issue.

From the documentation
```
cudaStreamCaptureModeThreadLocal: If the local thread has an ongoing capture sequence not initiated with cudaStreamCaptureModeRelaxed, it is prohibited from potentially unsafe API calls. Concurrent capture sequences in other threads are ignored.
``` my interpretation is `ThreadLocal` would still disallow the event query API call.

**Comment by bottler** [comment]
https://github.com/pytorch/pytorch/pull/103503 was a local change to the behaviour of a call on the ProcessGroup depending on the status of that operation's stream. It doesn't touch other streams or threads. 

I'm a bit confused about streams and threads here though. I would have thought that it's fine to have some "eager" threads with active nccl operations going on on some streams, and then to capture some nccl operations on a separate thread with a separate load of streams, and the capture shouldn't need to care about or wait for the eager threads and streams. I don't know if a single ProcessGroup object can allow that, but I don't see why not.

But maybe there's a problem with recycling streams which means we need to globally careful, i.e. a stream could be recycled from PyTorch's pool before the associate work on it has finished?

**Comment by eqy** [comment]
I think @kwen2501 raised an important point as to whether an operation being disallowed is the result of an actual underlying safety issue, or caused by the capture mode setting at the beginning of the capture. I observed that changing the capture mode to be thread local (or relaxed) also appears to resolve this issue, so it looks like it could also be the global capture mode simply disallowing any potentially unsafe API calls such as `cudaEventQuery`. I'll open a PR with that corresponding change and we can decide which route to take from there.

**Comment by bottler** [comment]
@eqy Can you share your repro code please? I'd like to concretise what the problem is. (Even better, but maybe tricky: could you add a test in this PR which would fail before the PR?)

**Comment by eqy** [comment]
I'm not sure I can publish the current repro as the code doesn't belong to me, but the internal asserts added are to check for that failure mode. The concrete issue is just that with capture mode global, any call to `cudaEventQuery` will crash the capture, as it is the strictest setting for CUDA graphs. In other words, if any process group had outstanding work that was not cleaned up by a watchdog before a graph capture, the watchdog can potentially crash the capture if it polls the work during the capture. Further info on capture modes: https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1g9d0535d93a214cbf126835257b16ba85

@Aidyn-A also raised an interesting point as to whether the thread local mode could potentially cause any problems with respect to autograd, where different threads will do the forward/backward passes.

**Comment by lw** [comment]
I agree with @kwen2501 and @eqy that a thread-local capture mode seems the easiest fix.

I want to also flag another potential fix, which has the benefit of not lowering the safety of graph capture: we could make the watchdog _catch_ any error that occurs when querying the event and retry if the error is `cudaErrorStreamCaptureUnsupported` (or something like that).

**Comment by bottler** [comment]
We can reproduce this bug easily now internally, with a workload which looks like this.

```python
for i in ...:
    for n_warmup in range(3):
          fn_with_allreduces(i)  # (on another stream)
    graph = <capture fn_with_allreduces(i)>
    for j in ...
         graph.replay()
```
and the crash happens in one of the later recordings. But the first recording seems reliably to work, I'm not sure why. Adding time.sleep(1) before capturing the graph fixes it.

cc @ruanslv

**Comment by eqy** [comment]
@bottler nice! The current proposed fix is # #105182 so it would be great if you could firm that it also addresses your repro.

**Comment by bottler** [comment]
> @bottler nice! The current proposed fix is #105182 so it would be great if you could firm that it also addresses your repro.

Effectively I can repro it by replacing the `with` [block](https://github.com/pytorch/pytorch/blob/94b3f9f646a84fb7bb0df997a57d410697440210/test/distributed/test_c10d_nccl.py#L443-L457) in  test_allreduce_in_cudagraph with

```
    with torch.cuda.device(rank):
        for i in range(100):
            xs = [torch.FloatTensor([1]).cuda(rank)]
            ys = [torch.FloatTensor([4]).cuda(rank)]
            for _ in range(30):
                pg.allreduce(xs[0]).wait()

            graph = torch.cuda.CUDAGraph()
            with torch.cuda.graph(graph):
                xs[0] += 0.0
                pg.allreduce(xs[0]).wait()
                pg.allreduce(xs[0]).wait()
                pg.allreduce(xs[0]).wait()
                xs[0] += 0.0

            for _ in range(1400):
                graph.replay()

```

And I can confirm that this doesn't crash if I make the changes in the latest version 9e44d98 of #105182 .

**Comment by eqy** [comment]
@bottler that's interesting, as 9e44d98 of https://github.com/pytorch/pytorch/pull/105182 is still setting the capture mode to global before doing an event query (which theoretically should crash). I was using this PR to temporarily test whether there is something wrong with using the exchange-capture-mode API, as in theory exchanging mode to global (the default) should basically be a NOP. However, it is causing the distributed tests to fail (at least on the M60 runners), so it looks like something else is going on.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp
Comment (kumpera) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:289]: This must be protected by a lock.

File: aten/src/ATen/cuda/CUDAGraph.cpp
Comment (kumpera) [inline aten/src/ATen/cuda/CUDAGraph.cpp:73]: We should never busyloop without sleeping.

**Review by kumpera — CHANGES_REQUESTED** [review]

**Comment by eqy** [comment]
CC @eellison @albanD 
Do you think we could consider this change as an alternative to the capture mode changes for the watchdog. Those distributed failures are a handful to debug, while this approach avoids changing capture modes entirely.

**Comment by eellison** [comment]
> This change as an alternative to the capture mode changes for the watchdog

is that an existing PR ? Sorry have been a number of PRs around NCCL/Cudagraphs. Also, in case you missed it, I am exposing the api to set cudaStreamCaptureMode in [another pr](https://github.com/pytorch/pytorch/pull/107407).

**Comment by eqy** [comment]
@eellison Yes, that PR is #104555 which was approved but unfortunately breaks distributed tests in a way that is very difficult to debug (seems to only surface on the specific M60 runners used, and doesn't appear to reproduce locally)

**Comment by eellison** [comment]
> Yes, that PR is https://github.com/pytorch/pytorch/pull/104555 which was approved but unfortunately breaks distributed tests in a way that is very difficult to debug

What about https://github.com/pytorch/pytorch/pull/105182, I thought that was the alternative to #104555 ? I guess 105182 had errors that weren't resolved ?

**Comment by eqy** [comment]
Yes, the errors were not resolved, and unfortunately it looks like it could be a bug specific to that specific generation of GPUs as I cannot reproduce the failure locally on Pascal, Ampere, etc.

File: aten/src/ATen/cuda/CUDAGraph.cpp
Comment (eellison) [inline aten/src/ATen/cuda/CUDAGraph.cpp:74]: printing this every 10ms seems like a lot

File: aten/src/ATen/cuda/CUDAGraph.cpp
Comment (eellison) [inline aten/src/ATen/cuda/CUDAGraph.cpp:80]: Don't we have a chance of a data race between these lines, if we start a new NCCL thread ? https://github.com/pytorch/pytorch/blob/36946bc6b5a5b9339c33f51a8446cdd2ba0391e0/aten/src/ATen/cuda/CUDAGraph.cpp#L80-L133

I don't know NCCL, so maybe this is not possible

Comment (eqy) [reply aten/src/ATen/cuda/CUDAGraph.cpp]: The graph capture begin is intended to be done on the main thread, and if someone is mixing side thread NCCL-usage and graph captures I believe they should be aware of the dangers at that point---there would also be a race also in that there would be no guarantees that their NCCL side thread work would be captured.

I've moved the check to be after the allocator's capture status is updated, just to keep things tidier.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp
Comment (kwen2501) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:717]: In case there are multiple threads and 1 GPU per thread, `all_nccl_process_groups` would include the PG objects on other threads. Not sure if this would create chance of data race (e.g. the return of `watchDogsDone` may depend on when _this_ thread check versus when the _other_ thread enqueues its collective.)

**Comment by kwen2501** [comment]
Apart from my minor comment above, the changes to ProcessGroupNCCL look okay to me.
I will defer to @eellison to make a call on whether mixing ProcessGroupNCCL into CUDAGraph.cpp is acceptable (hence final approval of this PR).
Longer term, I think it would be nice if we can go back to the `cudaGraphExchangeMode` solution in #105182.

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp
Comment (eqy) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:717]: The current solution is only intended to target the case where the work is enqueued by the main thread and captured on the main thread, as the failing case is work being enqueued during the _warmup_ on the main thread being queried during the capture itself. If work is enqueued on a side thread, then that is already unsafe as there would be no guarantee on what is captured/not captured, similar to Elias's concern.

File: aten/src/ATen/cuda/CUDAGraph.cpp
Comment (albanD) [inline aten/src/ATen/cuda/CUDAGraph.cpp:123]: Should we have a way to disable this? I could see some people not happy with the sleeps here?

File: aten/src/ATen/cuda/CUDAGraph.cpp
Comment (albanD) [inline aten/src/ATen/cuda/CUDAGraph.cpp:124]: This is going to be very spammy no? Should it only warn once?

File: aten/src/ATen/cuda/CUDAGraph.cpp
Comment (albanD) [inline aten/src/ATen/cuda/CUDAGraph.cpp:12]: Could you move the sleeping logic into ProcessGroupNCCL directly? It feels like having a `wait()` there makes sense.
Also side note: 10ms is an eternity for some user, not sure if that's ok?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp
Comment (albanD) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:289]: I assume someone did a careful review of the brand new locking here. I haven't looked in details.

File: aten/src/ATen/cuda/CUDAGraph.cpp
Comment (eqy) [reply aten/src/ATen/cuda/CUDAGraph.cpp:12]: The watchdogs already have sleeping logic, and the fact that wait loop occurs in a side thread (for fault-tolerance reasons) is why we need to wait to resynchronize here. 10ms was chosen here as it is the same polling interval used by the watchdog. I think it's acceptable to pay this as a one-time cost per graph capture---it would be a different story if this was needed for each iteration.

File: aten/src/ATen/cuda/CUDAGraph.cpp
Comment (eqy) [reply aten/src/ATen/cuda/CUDAGraph.cpp:123]: This codepath is already only enabled in graph captures, and I suspect in more than half the cases `watchDogsDone` will return `true`---the busy wait is more guarding against a somewhat rare but unlucky case that can crash things.

File: aten/src/ATen/cuda/CUDAGraph.cpp
Comment (albanD) [reply aten/src/ATen/cuda/CUDAGraph.cpp:12]: Right, this is generic logic to allow the current thread to wait on the watchdog thread to finish. My point is that such logic should live in the distributed code, not the cudagraph code.

Comment (eqy) [reply aten/src/ATen/cuda/CUDAGraph.cpp:12]: Do you have a recommendation for how to do this? Right now the function that does the check `watchdogsDone` is in `ProcessGroupNCCL`, but I don't see a clear reason why someone would want to poll this function outside of the graphs use-case.

Comment (albanD) [reply aten/src/ATen/cuda/CUDAGraph.cpp:12]: Anyone willing to check if the watchdog is done is most likely going to want to wait on it being done at some point?

**Review by albanD — COMMENTED** [review]
CUDAGraph side looks good.
I'll let someone from the distributed team to give a final stamp for the process group changes @kumpera @kwen2501

**Comment by kwen2501** [comment]
> CUDAGraph side looks good. I'll let someone from the distributed team to give a final stamp for the process group changes @kumpera @kwen2501

The changes on the distributed side look good to me, given that it is more urgent to unblock the functionality (CUDA Graph + Distributed). 

We will revisit #[105182](https://github.com/pytorch/pytorch/pull/105182) (graph mode change) whenever possible.

**Review by kwen2501 — APPROVED** [review]

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting you change, it is failing internal build' -c ghfirst

Here is the error FYI:

```
[2023-09-04T20:58:37.672-07:00] BUILD ERRORS (1)
[2023-09-04T20:58:37.672-07:00] The following actions failed during the execution of this command:
[2023-09-04T20:58:37.672-07:00] Action failed: fbcode//caffe2:ATen-cu (ovr_config//platform/linux:x86_64-fbcode-platform010-clang15-split-dwarf#363cd7e240f5d021) (cxx_compile aten/src/ATen/cuda/CUDAGraph.cpp)
[2023-09-04T20:58:37.672-07:00] Remote command returned non-zero exit code 1
[2023-09-04T20:58:37.672-07:00] Remote action, reproduce with: `frecli cas download-action dc7fdcbd27104997188af6cceca50f678bf453c695999f7349b079bae66cdc96:145`
[2023-09-04T20:58:37.672-07:00] Stdout: 
[2023-09-04T20:58:37.672-07:00] Stderr: fbcode/caffe2/aten/src/ATen/cuda/CUDAGraph.cpp:7:10: fatal error: 'torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp' file not found
[2023-09-04T20:58:37.672-07:00] #include <torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp>
[2023-09-04T20:58:37.672-07:00]          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[2023-09-04T20:58:37.672-07:00] 1 error generated.
[2023-09-04T20:58:37.672-07:00]
BUILD FAILED
```

The buck rule to build `CUDAGraph.cpp` would probably need to be updated.  It's probably a miss that the periodic buck build didn't catch this.

**Comment by eqy** [comment]
@huydhn Are there resources on how to test the buck build externally or are there internal changes/files required? e.g., https://github.com/pytorch/pytorch/blob/main/.github/workflows/_buck-build-test.yml looks like it requires Python 3.8.

**Comment by albanD** [comment]
@huydhn any update on this?

**Comment by huydhn** [comment]
Sorry for the delay.  I'm importing the pull request to verify it internally now.

**Comment by huydhn** [comment]
@albanD I have a diff to fix the internal issue here D49798453.  Please take a look to see if it's a sane solution.

File: aten/src/ATen/cuda/CUDAGraph.cpp
Comment (huydhn) [inline aten/src/ATen/cuda/CUDAGraph.cpp:7]: Another issue shows up.  Some internal Windows jobs are failing with the following error:

```
FAILED because xplat\caffe2\aten\src\ATen\cuda\CUDAGraph.cpp: included an untracked header: xplat\caffe2\torch\csrc\distributed\c10d\ProcessGroupNCCL.hpp
```

Comment (albanD) [reply aten/src/ATen/cuda/CUDAGraph.cpp:7]: The diff above should fix that I guess?

Comment (huydhn) [reply aten/src/ATen/cuda/CUDAGraph.cpp:7]: Well, the diff D49798453 fixes the first error when `ProcessGroupNCCL.hpp` isn't found.  Now, all the Linux builds are working, but the Windows builds are complaining about `ProcessGroupNCCL.hpp` is there but not tracked.  I think the new Windows build failure could be fixed here by adding `#ifdef USE_C10D_NCCL` or `#ifndef _WIN32` (NCCL is not available on Windows).  Then it would be good to go. @eqy Could you help update the PR for that?

**Pull request closed without merging**
